akka-actor-typed-tests/src/test/scala/akka/actor/typed/SupervisionSpec.scala

```diff
@@ -548,6 +552,50 @@ class SupervisionSpec extends ScalaTestWithActorTestKit("""
       parentProbe.expectMessage(ReceivedSignal(Terminated(anotherProbe.ref)))
     }
 
+    "successfully restart after stopping watchWith'd children" in {
+      val parentProbe = TestProbe[Event]("evt")
+      val behv = Behaviors.supervise(targetBehavior(parentProbe.ref)).onFailure[Exc1](SupervisorStrategy.restart)
+      val ref = spawn(behv)
+
+      val anotherProbe = TestProbe[String]("another")
+      ref ! WatchWith(anotherProbe.ref, Ping(0))
+
+      val childProbe = TestProbe[Event]("childEvt")
+      val slowStop = new CountDownLatch(1)
+      val childName = nextName()
+      ref ! CreateChild(targetBehavior(childProbe.ref, slowStop = Some(slowStop)), childName)
+      ref ! GetState
+      val childRef =
+        parentProbe.receiveMessage() match {
+          case State(0, children) =>
+            children.keySet should ===(Set(childName))
+            children(childName)
+          case _ =>
+            fail("expected to receive a State(0, _)")
+        }
+
+      ref ! WatchWith(childRef, Ping(1))
+
+      LoggingTestKit.error[Exc1].expect {
+        ref ! Throw(new Exc1)
+        parentProbe.expectMessage(ReceivedSignal(PreRestart))
+        ref ! GetState
+        anotherProbe.stop()
+      }
+
+      // waiting for children to stop, GetState stashed
+      parentProbe.expectNoMessage()
+      slowStop.countDown()
+
+      childProbe.expectMessage(ReceivedSignal(PostStop))
+      parentProbe.expectMessageType[State].children.keySet should ===(Set.empty)
+      // we get the Ping(0) message from stopping anotherProbe
+      parentProbe.expectMessage(Pong(0))
+      // but we didn't get the Ping(1) message from stopping the child, because the
+      // restart strategy revokes the watchWith in favor of watch
```

Comment:
The semantics around not sending the `watchWith` message to the restarted actor match the semantics for the `Terminated` signal in the `watch` case: it was handled by the restart supervisor and since the number of messages/signals arising from `watch`/`watchWith` can never exceed the number of invocations, it can't be passed on.